src/data/markdown/docs/01 guides/05 Testing Guides/04 Large-scale k6 testing.md

```diff
@@ -0,0 +1,484 @@
+When the system starts to fail, the above check won't work as expected. 
+
+The issue here is that the check assumes that there's  always a body in a response. The `r.body` may not exist if server is failing.
+In such case the check itself won't work as expected and error similar to the one below will be returned: 
+
+<div class="code-group" data-props='{}'>
+
+```
+ERRO[0625] TypeError: Cannot read property 'length' of undefined
+```
+
+</div> 
+
+To fix this issue your checks must be resilient to any response type. This change will fix the above problem.
+
+<div class="code-group" data-props='{"labels": ["resilient check"]}'>
+
+```javascript
+let checkRes = check(res, {
+    "Homepage body size is 11026 bytes": (r) => r.body && r.body.length === 11026
+});
+```
+
+</div>
+
+### Monitor the load-generator server
+
+If you are running a test for the first time, it's a good idea to keep an eye on the available resources while the test is running.
+The easiest way to do so is to ssh to the server with 3 sessions:
+1. To run k6 
+2. To monitor CPU and memory
+3. To monitor the network
+
+For monitoring CPU and memory we recommend [htop](https://en.wikipedia.org/wiki/Htop). 
+For monitoring network, we recommend [iftop](https://en.wikipedia.org/wiki/Iftop)
+
+Here's a screenshot of 3 terminal sessions showing k6, iftop and htop.
+![k6 iftop htop](./images/large-scale-testing-3-terminals.png)
+
+
+## Additional k6 flags to achieve better performance
+
+### --compatibility-mode=base
+
+If you are pushing the limits of the hardware, this is the most impactful k6 setting you can enable.
+
+This setting disables the internal babel transpilation from ES6 to ES5 and inclusion of corejs library.
+
+<div class="doc-blockquote" data-props='{}'>
+
+> ### Some background
+> k6 at its core executes ECMAScript 5.1 code. Most k6 script examples and documentation is written in ECMAScript 6+. 
+> When k6 gets ES6+ code as input, it transpiles it to ES5.1 using babel and loads corejs to enable commonly used APIs.
+> This works very well for 99% of use cases, but it adds significant overheard with large tests.
+
+</div>
+
+To get the best performance out of k6, it's best to transpile the scripts outside of k6 using webpack.
+
+In [k6-hardware-benchmark](https://github.com/loadimpact/k6-hardware-benchmark) repository, we have prepared an efficient transpilation scheme that produces performant ES5.1 code for k6. 
+
+Use it like this:
+
+<div class="code-group">
+
+```bash
+git clone https://github.com/loadimpact/k6-hardware-benchmark/
+cd k6-hardware-benchmark
+yarn install
+yarn run to-es5 someplace/yourscript.js
+# your ES5 script is in someplace/yourscript.es5.js
+```
+</div>
+
+
+Once your code is transpiled, run it like this:
+
+<div class="code-group">
+
+```bash
+k6 run -o cloud --compatibility-mode=base someplace/yourscript.es5.js
+```
+</div>
+
+k6 will use about 50% of memory in comparison to running the original script. It will also reduce the CPU load. 
```

Comment:
I would argue that if your script is simple and uses discardResponseBodies the difference is more like [85%](https://github.com/loadimpact/k6/issues/1167#issuecomment-553835092) , with a 1.5mb json data file(If my memory served) the difference was 50%, but I can't currently find where I've said that :sob: . I would argue some links to benchmark data will be useful here.

Also another thing that compatibility-mode=base does is "significantly" increase the startup time. In the above linked script I ran the script for 60s and in the one case there were additiona minute and 5 seconds (on average) additiona time in the other a secodn and a half for 3500 VUs. Which IMO is "significant" :D

Reply:
Updated the text. If you have relevant benchmark links, let me know.